### PR TITLE
Add an example data pipeline to transfer data based on MSSQL

### LINKED ISSUE #13: Setup an example data pipeline task (container) for data transfer using pyodbc - mssql
Setup an example data pipeline task (container) for data transfer using pyodbc.

Requirements:
 - Add an example project - `examples/data-transfer-mssql`
 - Have two mssql databases - source and destination; and transfer data from first database's table to the destination database's table.

### PR DESCRIPTION
**Pull request by razzkumar**
**Changes**
New example added which create two mssql server and select data from one db to insert to another db
final-final

Resolves #13

### CODE REVIEW

File: examples/app-mssql/main.py [hunk 1]
```diff
@@ -22,7 +22,7 @@ def main():
     conn = pyodbc.connect(connection_str, timeout=300)
     cur = conn.cursor()
 
-    print('Create a new table for fruits.')
+    print('Create a new table for fruits. idhf saiufhdsauifh iusdafhisd faiu h')
```
Comment (kabirbaidhya) [inline examples/app-mssql/main.py]: What is this bro?

File: examples/app-mssql/main.py [hunk 2]
```diff
@@ -32,7 +32,7 @@ def main():
     cur.execute('INSERT INTO fruits VALUES (3, ?, ?)', ('Apple', 35))
     conn.commit()
 
-    print('List of data.')
+    print('List of data.--------------------------------')
```
Comment (kabirbaidhya) [inline examples/app-mssql/main.py]: Undo.

File: examples/data-transfer-mssql/README.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
+# MSSQL Example
+
+Example project using mssql with pyodbc docker image. which take a data from
+one db and insert to another table
```
Comment (kabirbaidhya) [inline examples/data-transfer-mssql/README.md]: Example project to demonstrate a data pipeline container built using `laudio/pyodbc` as a base image - that transfers data from one database to another database.

File: examples/data-transfer-mssql/README.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
+# MSSQL Example
```
Comment (kabirbaidhya) [inline examples/data-transfer-mssql/README.md]: Data Transfer for MSSQL

File: examples/data-transfer-mssql/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+  app:
+    build: .
+    depends_on:
+      - mssql_source
+      - mssql_destination
+    environment:
+      SOURCE_DB_HOST: mssql_source
+      DESTINATION_DB_HOST: mssql_destination
+      DB_NAME: ${DB_NAME}
```
Comment (kabirbaidhya) [inline examples/data-transfer-mssql/docker-compose.yml]: Have different database names for source and destination.

File: examples/data-transfer-mssql/main.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        server=os.environ['DESTINATION_DB_HOST'],
+        database=os.environ['DB_NAME'],
+        username=os.environ['DB_USERNAME'],
+        password=os.environ['DB_PASSWORD']
+    )
+
+    destination_conn = pyodbc.connect(destination_connection_str, timeout=300)
+    cur2 = destination_conn.cursor()
+
+    print('\n\n\nCreate a new table for fruits. on destination db')
```
Comment (kabirbaidhya) [inline examples/data-transfer-mssql/main.py]: Create table on the destination database.

File: examples/data-transfer-mssql/main.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+    cur2.execute('CREATE TABLE fruits (id INT, name NVARCHAR(50), quantity INT)')
+    destination_conn.commit()
+
+    print('Inserting data to destination db')
```
Comment (kabirbaidhya) [inline examples/data-transfer-mssql/main.py]: Transferring data into destination database.

File: examples/data-transfer-mssql/main.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    for row_data in source_rows:
+        cur2.execute('INSERT INTO fruits VALUES (?, ?, ?)', (row_data.id,row_data.name,row_data.quantity))
+        destination_conn.commit()
+
+    cur2.execute('SELECT * FROM fruits')
+    destination_rows = cur2.fetchall()
+
+    print('----------------List of data from destination.-----------------------')
```
Comment (kabirbaidhya) [inline examples/data-transfer-mssql/main.py]: Data in the destination database.

File: examples/app-mssql/main.py (hunk 1 above)
Comment (mesaugat) [reply examples/app-mssql/main.py]: LOL.

File: examples/app-mssql/main.py (hunk 2 above)
Comment (mesaugat) [reply examples/app-mssql/main.py]: Gone

File: examples/data-transfer-mssql/main.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    cur1.execute('INSERT INTO fruits VALUES (1, ?, ?)', ('Banana', 150))
+    cur1.execute('INSERT INTO fruits VALUES (2, ?, ?)', ('Orange', 64))
+    cur1.execute('INSERT INTO fruits VALUES (3, ?, ?)', ('Apple', 35))
+    source_conn.commit()
+
+    print('---selecting data from source db----')
+    cur1.execute('SELECT * FROM fruits')
+    source_rows = cur1.fetchall()
+    
+    print('---closing source db----')
```
Comment (razzkumar) [inline examples/data-transfer-mssql/main.py]: Closing source database connection.

**Comment by kabirbaidhya** [comment]
@razzkumar Thanks for contributing brother. Please make sure you make everything (except mssql specifics) identical to this PR https://github.com/laudio/pyodbc/pull/19 as except the tech stack everything is similar. And also stay tuned to the comments there - more is coming :grin: :crossed_fingers:

**Comment by razzkumar** [comment]
@kabirbaidhya dai can you please briefly explain me about the transferring data by sending flag while running docker without making database to transfer existing data, currently, I have created a database in source and insert the data into a table and transfer the data by retrieving inserted data

File: examples/data-transfer-mssql/sql/source_db_setup.sql [hunk 10]
```diff
@@ -0,0 +1,7 @@
+-- Create table for fruits data (source database).
+CREATE TABLE fruits (id INT,name VARCHAR(50),quantity INT);
```
Comment (mesaugat) [inline examples/data-transfer-mssql/sql/source_db_setup.sql]: Space after comma.

```suggestion
CREATE TABLE fruits (id INT, name VARCHAR(50), quantity INT);
```

File: examples/data-transfer-mssql/sql/dest_db_setup.sql [hunk 11]
```diff
@@ -0,0 +1,2 @@
+-- Create table for fruits data (destination database).
+CREATE TABLE fruits (id INT,name VARCHAR(50),quantity INT);
```
Comment (mesaugat) [inline examples/data-transfer-mssql/sql/dest_db_setup.sql]: Space after a comma, add a new line at EOF.

File: examples/data-transfer-mssql/sql/source_db_setup.sql [hunk 12]
```diff
@@ -0,0 +1,7 @@
...
+
+-- Insert data into fruits table
+INSERT INTO fruits VALUES (1, 'Banana', 200);
+INSERT INTO fruits VALUES (2, 'Orange', 20);
+INSERT INTO fruits VALUES (3, 'Apple', 350);
```
Comment (mesaugat) [inline examples/data-transfer-mssql/sql/source_db_setup.sql]: New line at EOF.

Add this setting if you are using VSCode.

```
"files.insertFinalNewline": true
```

File: examples/data-transfer-mssql/main.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+            os.environ['DESTINATION_DB_HOST'],
+            os.environ['DESTINATION_DB_NAME'],
+            os.environ['DESTINATION_DB_USERNAME'],
+            os.environ['DESTINATION_DB_PASSWORD']
+        )
+    
+    return source_db_conn, dest_db_conn
+
+def get_connection(db_host,db_name,db_username,db_password):
+
```
Comment (mesaugat) [inline examples/data-transfer-mssql/main.py:53]: Missing doc block?

```
'''
```

File: examples/data-transfer-mssql/README.md [hunk 14]
```diff
@@ -0,0 +1,43 @@
+# MSSQL Data Transfer Example
+
+Example project to demonstrate a data pipeline container built using laudio/pyodbc as a base image - that transfers data from one mssql database to another mssql database
```
Comment (kabirbaidhya) [inline examples/data-transfer-mssql/README.md]: Full stop. 
```suggestion
Example project to demonstrate a data pipeline container built using laudio/pyodbc as a base image - that transfers data from one mssql database to another mssql database.
```

File: examples/data-transfer-mssql/main.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+def extract_sql(file: str):
+    ''' Reads an SQL file and returns it's contents.'''
+    with open(file, 'rt') as file:
+        contents = file.read()
+    return contents
+
+
+def transfer_data(source_db_cursor, dest_db_cursor, dest_db_conn):
+    ''' Extracts fruits data from source database and stores them in destination database.'''
```
Comment (kabirbaidhya) [inline examples/data-transfer-mssql/main.py]:
```suggestion
    ''' Extracts fruits data from source database and stores them in destination database. '''
```

File: examples/data-transfer-mssql/main.py [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+            dest_db_conn.commit()
+
+            transfer_data(source_db_cur, dest_db_cur, dest_db_conn)
+
+            print(f'Display fruits data of destination database.')
+            display_fruits(dest_db_cur)
+
+
+def connect_to_databases():
+    ''' Extracts databases credentials from the environment and returns their connections.'''
```
Comment (kabirbaidhya) [inline examples/data-transfer-mssql/main.py]:
```suggestion
    ''' Extracts databases credentials from the environment and returns their connections. '''
```

**Comment by kabirbaidhya** [comment]
> @kabirbaidhya dai can you please briefly explain me about the transferring data by sending flag while running docker without making database to transfer existing data, currently, I have created a database in source and insert the data into a table and transfer the data by retrieving inserted data

@razzkumar For this PR it's OK as long as it's similar to the pg example. 
I meant we need a way or a  "flag" to decide whether or not to setup database / tables. Because we only need to create them once - so that when we run it the second time it'd only be data transfer w/o setup. 

But this could be done in a later PR. For this one just fix few of the minor comments mentioned above.

File: examples/data-transfer-mssql/main.py [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+
+def main():
+    ''' App entrypoint. '''
+    # Wait for mssql database server to fully spawn.
+    time.sleep(5)
+
+    source_db_conn, dest_db_conn = connect_to_databases()
+
+    with source_db_conn, dest_db_conn:
+
```
Comment (mesaugat) [inline examples/data-transfer-mssql/main.py]: Do we need a new line here? cc @kabirbaidhya

File: examples/data-transfer-mssql/main.py [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+        source_db_cur = source_db_conn.cursor()
+        dest_db_cur = dest_db_conn.cursor()
+
+        with source_db_cur, dest_db_cur:
+
```
Comment (mesaugat) [inline examples/data-transfer-mssql/main.py]: same

File: examples/data-transfer-mssql/main.py [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+            os.environ['DESTINATION_DB_HOST'],
+            os.environ['DESTINATION_DB_NAME'],
+            os.environ['DESTINATION_DB_USERNAME'],
+            os.environ['DESTINATION_DB_PASSWORD']
+        )
+    
+    return source_db_conn, dest_db_conn
+
+def get_connection(db_host,db_name,db_username,db_password):
+    ''' create database connection and returns connection. '''
```
Comment (mesaugat) [inline examples/data-transfer-mssql/main.py]:
```diff
- create
+ Create
```

File: examples/data-transfer-mssql/main.py [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+        database=db_name,
+        username=db_username,
+        password=db_password
+    )
+    
+    return pyodbc.connect(connection_str, timeout=300)
+
+
+def extract_sql(file: str):
+    ''' Reads an SQL file and returns it's contents.'''
```
Comment (mesaugat) [inline examples/data-transfer-mssql/main.py]:
```suggestion
    ''' Reads an SQL file and returns it's contents. '''
```

File: examples/data-transfer-mssql/README.md [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+$ cp .env.example .env
+```
+
+Run the example.
+
+```bash
+$ docker-compose up
+```
+
+For changes
```
Comment (mesaugat) [inline examples/data-transfer-mssql/README.md]: Do we need this block? Not sure we need this information. cc @kabirbaidhya

File: examples/data-transfer-mssql/main.py [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+            transfer_data(source_db_cur, dest_db_cur, dest_db_conn)
+
+            print(f'Display fruits data of destination database.')
+            display_fruits(dest_db_cur)
+
+
+def connect_to_databases():
+    ''' Extracts databases credentials from the environment and returns their connections.'''
+    source_db_conn = get_connection(
+            os.environ['SOURCE_DB_HOST'],
```
Comment (mesaugat) [inline examples/data-transfer-mssql/main.py]: Formatting here, use 4 spaces?

```python
source_db_conn = get_connection(
    os.environ['SOURCE_DB_HOST'],
    os.environ['SOURCE_DB_NAME'],
    os.environ['SOURCE_DB_USERNAME'],
    os.environ['SOURCE_DB_PASSWORD']
)

File: examples/data-transfer-mssql/main.py [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+            os.environ['SOURCE_DB_NAME'],
+            os.environ['SOURCE_DB_USERNAME'],
+            os.environ['SOURCE_DB_PASSWORD']
+        )
+
+    dest_db_conn = get_connection(
+            os.environ['DESTINATION_DB_HOST'],
+            os.environ['DESTINATION_DB_NAME'],
```
Comment (mesaugat) [inline examples/data-transfer-mssql/main.py]: Same.

**Comment by razzkumar** [comment]
@kabirbaidhya and  @mesaugat dai I am really thankful for reviews and suggestions. As this is my first opensource pull request (probably 6th or 7th pull request of my life), I have learned a lot.

**Review by mesaugat — APPROVED** [review]
Awesome. Looks good to me brother. :100: @kabirbaidhya will handle the rest.

File: examples/data-transfer-mssql/main.py [hunk 24]
```diff
@@ -0,0 +1,104 @@
+''' Main module. '''
```
Comment (mesaugat) [inline examples/data-transfer-mssql/main.py]: We don't need this comment here.

**Review by kabirbaidhya — APPROVED** [review]
Awesome! Thanks for contributing @razzkumar :100: :+1:. And keep contributing to opensource and you'll surely learn a lot more moving forward.

**Pull request merged**